Lean cleanup DAG tutorial (#3641)
astronomer/docs: learn/cleanup-dag-tutorial.md

---
title: 'Clean up the Airflow metadata database using DAGs'
sidebar_label: 'Clean up the metadata database'
id: cleanup-dag-tutorial
description: 'Learn how to remove unnecessary data from the Airflow metadata database using the `airflow db clean` command from a DAG.'
---

import CodeBlock from '@theme/CodeBlock';
import db_cleanup from '!!raw-loader!../code-samples/dags/cleanup-dag-tutorial/db_cleanup.py';

> In addition to storing configuration about your Airflow environment, the Airflow [metadata database](https://docs.astronomer.io/learn/airflow-database) stores data about past and present task runs. The more you use Airflow, the more data the metadata DB stores about these task runs. Over a long enough time, this can result in a bloated metadata DB, which can affect performance across your Airflow environment.

Reviewer: ```suggestion
In addition to storing configuration about your Airflow environment, the Airflow [metadata database](https://docs.astronomer.io/learn/airflow-database) stores data about past and present task runs. Airflow never automatically removes metadata, so the longer you use it, the more task run data is stored in your metadata DB. Over a long enough time, this can result in a bloated metadata DB, which can affect performance across your Airflow environment.
```

> When a table in the metadata DB is larger than 50GB, you might start to experience degraded scheduler performance. This can result in:

Author: Is there a more general figure that we could use here for all Airflow environments? This is when the alert triggers on Astro, but I imagine it can start to get slower a lot quicker for other enviros. What figure can OSS users check to see if a table is getting too big?

- Slow task scheduling
- Slow DAG parsing

> - [List more specifics]

Author: Any other side effects we should mention here?

Reviewer 2: I can think of two web server issues:
1. Gunicorn timeout when a query takes a longer time to complete. 
2. Slower UI loading

The following tables in the database are at risk of becoming too large over time:

- `job` 
- `dag_run` 
- `task_instance`
- `log` 
- `dags` 
- `task_reschedule`
- `task_fail`

> To keep your Airflow environment running at optimal performance, you can clean the metadata DB using the `airflow db clean` command. This tutorial walks you through how to implement a cleanup DAG in Airflow so that you can clean your database using the command directly from the Airflow UI.

Reviewer: ```suggestion
To keep your Airflow environment running at optimal performance, you can clean the metadata DB using the Airflow CLI `airflow db clean` command. This command was created as a way to safely clean up your metadata DB; you should never modify the Airflow metadata DB directly, as this can cause dependency issues and corrupt your Airflow instance. This tutorial walks you through how to implement a cleanup DAG in Airflow so that you can clean your database using the command directly from the Airflow UI.
```
Feel free to adjust where we put this, but I think the extra context is helpful.

## Warnings

:::danger

> Read this section carefully before implementing this DAG in any production Airflow environments.

Reviewer: I almost wonder if we should move this up into the intro and update to "Read the Warnings section....", in case anybody happens to scroll over it (not that they couldn't do that anyway, but maybe this would help)

Author: I moved this slightly higher, but not all the way to the beginning so as to avoid https://en.wikipedia.org/wiki/Banner_blindness

:::

Deleting data from the metadata database can be an extremely destructive action. If you delete data that future task runs depend on, it's difficult to recover the database to its previous state without interrupting your data pipelines. Before implementing the DAG in this tutorial, consider the following:

> - Only trigger a database cleanup if you need to. If your Airflow scheduler is running normally, there's no need to clean the DB. On Astro, you will receive alerts when specific tables contain enough data to cause performance issues.

Author: More usage warnings here would be great, these are just the ones I can think of off the top of my head

- Delete older data before newer data. The older the deleted data, the less likely it is to affect your currently running DAGs.

> - The DAG will fail if it runs for longer than five minutes, which can happen when you have an exceptionally large amount of data to delete. If this happens to you, use the DAG's params to reduce the amount of data being deleted at once, either by reducing the time window for deletion or reducing the number of tables to clean.

Reviewer: Is this true generally, or only on Astro? I think Felix said the latter, but we should clarify.

## Prerequisites

- An Astro project
- The Astro CLI

> ## Step 1: Create your DAG

Reviewer: Somewhere (in this section? the intro?) we should add a note that this DAG has been optimized for Astro users, and while it's still relevant for OSS users, they may want to review and change some of the defaults.

Author: Is it though? By default it deletes the data in all tables older than 90 days, which doesn't seem too Astro specific to me, but I might be missing some context here

Reviewer 3: I think this is the answer to your question above about 50 GB. That's difficult for us to know in other environments, so just state this is for Astro

1. In your dags folder, create a file called `db_cleanup.py`.

2. Copy the following code into the file.

    <CodeBlock language="python">{db_cleanup}</CodeBlock>

    Rather than running on a schedule, this DAG is triggered manually by default and includes params so that you're in full control over how you clean the metadata DB.

>     It includes one task, `clean_db`, that runs the `airflow db clean` with the params you specify at Runtime. The params let you specify:

Reviewer: ```suggestion
    It includes one task, `clean_db`, that runs `airflow db clean` with the params you specify at Runtime. The params let you specify:
```

>     - What age of data to delete. Any data that was created before the specified time will be deleted. 

Reviewer: I think in all of these bullets worth stating what the default is, in case people don't read the DAG code

>     - Whether to run the cleanup as a dry run, meaning that no data is deleted.

Reviewer: ```suggestion
    - Whether to run the cleanup as a dry run, meaning that no data is deleted. The DAG will return the SQL that would be executed based on other parameters you have specified.
```

    - Which tables to delete data from.

> ## Step 2: Run your DAG

Reviewer: It seems like this would be more useful if the step was to deploy the DAG to astro and run as needed? Or if we want to explicitly tell people to run locally so they can "practice" with it, maybe we should state that.

- Slow DAG parsing
- [List more specifics]

The following tables in the database are at risk of becoming too large over time:

- `dag_run` 
- `job` 
- `log` 
- `rendered_task_instance_fields`
- `task_instance`
- `xcom`

To keep your Airflow environment running at optimal performance, you can clean the metadata DB using the Airflow CLI `airflow db clean` command. This command was created as a way to safely clean up your metadata DB without querying it directly. This tutorial walks you through how to implement a cleanup DAG in Airflow so that you can clean your database using the command directly from the Airflow UI.

:::danger

Even when using `airflow db clean`, deleting data from the metadata database can be destructive . Read the [Warnings](#warnings) section carefully before implementing this tutorial DAG in any production Airflow environments.

:::

## Warnings

Deleting data from the metadata database can be an extremely destructive action. If you delete data that future task runs depend on, it's difficult to recover the database to its previous state without interrupting your data pipelines. Before implementing the DAG in this tutorial, consider the following:

> - Delete older data before newer data. The older the deleted data, the less likely it is to affect your currently running DAGs.

Reviewer 3: It's not possible to delete new data before old data with this DAG

Reviewer 4: I think what we want to say is to start by deleting very old records. For example, start with older than 180 days, before deleting data older than 90 days.

Reviewer 3: I would state that directly then. This sentence strongly implies that it's possible to end up with a hole in your data.

> - On Astro, the DAG will fail if it runs for longer than five minutes, which can happen if you have an exceptionally large amount of data to delete. If this happens, use the DAG's params to reduce the amount of data being deleted at once, either by reducing the time window for deletion or reducing the number of tables to clean.

Reviewer 3: Reducing the number of tables won't help. The timeout is per table.

Reviewer 4: That is correct. The timeout is per query, and since the cleanup task runs a (cleanup) query per table it won't help.

## Prerequisites

> - An Astro project

Reviewer 3: An Airflow project, nothing too Astro specific here

Author: Our standard for tutorials is to say "Astro project", since we instruct users to run the DAG with the Astro CLI and you can't use an OSS Airflow project for that.

sidebar_label: 'Clean up the metadata database'
id: cleanup-dag-tutorial
description: 'Learn how to remove unnecessary data from the Airflow metadata database using the `airflow db clean` command from a DAG.'
---

import CodeBlock from '@theme/CodeBlock';
import db_cleanup from '!!raw-loader!../code-samples/dags/cleanup-dag-tutorial/db_cleanup.py';

In addition to storing configuration about your Airflow environment, the Airflow [metadata database](https://docs.astronomer.io/learn/airflow-database) stores data about past and present task runs. Airflow never automatically removes metadata, so the longer you use it, the more task run data is stored in your metadata DB. Over a long enough time, this can result in a bloated metadata DB, which can affect performance across your Airflow environment.

When a table in the metadata DB is larger than 50GB, you might start to experience degraded scheduler performance. This can result in:

- Slow task scheduling
- Slow DAG parsing
- [List more specifics]

The following tables in the database are at risk of becoming too large over time:

- `job` 
- `dag_run` 
- `task_instance`
- `log` 
- `dags` 
- `task_reschedule`

> - `task_fail`

Reviewer 5: xcom, rendered_task_instance_fields, more so than task_reschedule and task_fail. I don't think dags is a risk either.

The order feels a bit odd to me also.

> To keep your Airflow environment running at optimal performance, you can clean the metadata DB using the Airflow CLI `airflow db clean` command. This command was created as a way to safely clean up your metadata DB without modifying it directly, as this can cause dependency issues and corrupt your Airflow instance. This tutorial walks you through how to implement a cleanup DAG in Airflow so that you can clean your database using the command directly from the Airflow UI.

Reviewer 5: ```suggestion
To keep your Airflow environment running at optimal performance, you can clean the metadata DB using the Airflow CLI `airflow db clean` command. This command was created as a way to safely clean up your metadata DB without querying it directly. This tutorial walks you through how to implement a cleanup DAG in Airflow so that you can clean your database using the command directly from the Airflow UI.
```

"querying" vs "modifying", and I'd drop the reason since FKs in theory should avoid corruption.

:::danger

Even when using `airflow db clean`, deleting data from the metadata database can be destructive . Read the [Warnings](#warnings) section carefully before implementing this tutorial DAG in any production Airflow environments.

:::

## Warnings

Deleting data from the metadata database can be an extremely destructive action. If you delete data that future task runs depend on, it's difficult to recover the database to its previous state without interrupting your data pipelines. Before implementing the DAG in this tutorial, consider the following:

> - Only trigger a database cleanup if you need to. If your Airflow scheduler is running normally, there's no need to clean the DB. On Astro, you will receive alerts when specific tables contain enough data to cause performance issues.

Reviewer 5: I'm not a fan of saying "no need unless there is pain". It is good hygiene and ideally folks do it proactively.

- Slower Airflow UI load times

The following tables in the database are at risk of becoming too large over time:

- `dag_run` 
- `job` 
- `log` 
- `rendered_task_instance_fields`
- `task_instance`
- `xcom`

To keep your Airflow environment running at optimal performance, you can clean the metadata DB using the Airflow CLI `airflow db clean` command. This command was created as a way to safely clean up your metadata DB without querying it directly. This tutorial walks you through how to implement a cleanup DAG in Airflow so that you can clean your database using the command directly from the Airflow UI.

:::danger

Even when using `airflow db clean`, deleting data from the metadata database can be destructive . Read the [Warnings](#warnings) section carefully before implementing this tutorial DAG in any production Airflow environments.

:::

## Warnings

Deleting data from the metadata database can be an extremely destructive action. If you delete data that future task runs depend on, it's difficult to recover the database to its previous state without interrupting your data pipelines. Before implementing the DAG in this tutorial, consider the following:

> - When specifying the `clean_before_timestamp` value, use as late of a data as possible. The older the deleted data, the less likely it is to affect your currently running DAGs.

Author: Typo on "data" here, but I think this should still be the "latest date". If you select a recent date, it will delete all data before the recent date, when we specifically want users only to delete later data if possible.

> - When specifying the `clean_before_timestamp` value, use as late of a data as possible. The older the deleted data, the less likely it is to affect your currently running DAGs.

Reviewer 6: Wouldn't that be the "oldest date" then? When I read "latest date" to me that translates to "most recent," not the "most distant in the past"?

> - When specifying the `clean_before_timestamp` value, use as late of a date as possible. The older the deleted data, the less likely it is to affect your currently running DAGs.

Reviewer 6: ```suggestion
- When specifying the `clean_before_timestamp` value, use as old a date as possible. The older the deleted data, the less likely it is to affect your currently running DAGs.
```

> - On Astro, the DAG will fail if it runs for longer than five minutes, which can happen if you have an exceptionally large amount of data to delete. If this happens, use the DAG's params to reduce the amount of data being deleted from a single table at once by moving the `clean_before_timestamp` earlier.

Author: Same here, a more recent time will cause more data to be deleted.

## Step 1: Create your DAG

:::warning

This DAG has been designed and optimized for Airflow environments running on Astro. Consider adjusting the parameters and code if you're running the DAG in any other type of Airflow environment.

:::

1. In your dags folder, create a file called `db_cleanup.py`.

2. Copy the following code into the file.

    <CodeBlock language="python">{db_cleanup}</CodeBlock>

    Rather than running on a schedule, this DAG is triggered manually by default and includes params so that you're in full control over how you clean the metadata DB.

    It includes one task, `clean_db`, that runs `airflow db clean` with the params you specify at runtime. The params let you specify:

    - What age of data to delete. Any data that was created before the specified time will be deleted. The default is to delete all data older than 90 days.
    - Whether to run the cleanup as a dry run, meaning that no data is deleted. The DAG will instead return the SQL that would be executed based on other parameters you have specified. The default is to run the deletion without a dry run.
    - Which tables to delete data from. The default is all tables.

## Step 2: Run the DAG

> In this step, you'll run the DAG in a local Airflow environment to practice the workflow for cleaning metadata DB data. In reality, you would first deploy this DAG to a production environment, such as an Astro Deployment, then run the DAG only when your metadata DB was getting full.

Reviewer 6: ```suggestion
In this step, run the DAG in a local Airflow environment to practice the workflow for cleaning metadata DB data. When using this process for a production environment, you would first deploy this DAG to a production environment, such as an Astro Deployment, then run the DAG only when your metadata DB is getting full.
```

1. Run `astro dev start` in your Astro project to start Airflow, then open the Airflow UI at `localhost:8080`.

2. In the Airflow UI, run the `db_cleanup` DAG by clicking the play button, then click **Trigger DAG w/ Config**. Configure the following params:

    - `dry_run`: `true`
    - `tables`: `all_tables`
    - `clean_before_timestamp`: `datetime.now(tz=UTC) - timedelta(days=90)`

3. Click **Trigger**. 
4. After the task completes, click **Graph**.
5. Click a task run.
6. Click **Instance Details**.
7. Click **Log**.

> 8. Check that the `airflow db cleanup` command completed successfully. Note that you created a new Astro project, the run will not show much data to be deleted. 

Reviewer 6: ```suggestion
8. Check that the `airflow db cleanup` command completed successfully. Note that if you created a new Astro project, the run will not show much data to be deleted. 
```
